### PR TITLE
Added tests for local/remote bindings for deduper/mixer; tests in CI

### PR DESCRIPTION
**Pull request by soldni**

### CODE REVIEW

File: tests/config/paragraph-spans.json [hunk 1]
```diff
@@ -3,10 +3,10 @@
     {
       "name": "paragraph-spans-test",
       "documents": [
-        "s3://ai2-llm/pretraining-data/tests/mixer/inputs/v0/documents/*/0000.json.gz"
+        "tests/data/provided/documents/*"
```
Comment (rodneykinney) [inline tests/config/paragraph-spans.json:6]: I'm a little confused. Are the tests reading writing to S3? Some of them? I see some S3 uploading happening in the Python `TestCase`s.

My preference would be to have most of the tests operating on purely local data, to fully test the logic in the mixer and deduper. There should be at least one test that reads/writes from S3, just to ensure that's working, but it doesn't have to verify any internal logic.

**Review by rodneykinney — COMMENTED** [review]
Nice! And a lot of work. Some mostly-nit-level comments.

File: tests/config/paragraph-spans.json (hunk 1 above)
Comment (soldni) [reply tests/config/paragraph-spans.json:6]: all tests run with local file by default; further, `test_dedupe_by_url_remote_input`, ` test_local_input_remote_output`, `test_remote_input_remote_output`, and `test_remote_input_local_output` run on a combination of local/remote input/output.

File: src/shard.rs [hunk 2]
```diff
@@ -581,7 +603,7 @@ pub fn find_objects_matching_patterns(patterns: &Vec<String>) -> Result<Vec<Stri
         let mut matches = Vec::new();
         for pattern in patterns.iter() {
             for entry in
-            glob(pattern).expect(format! {"Invalid file pattern: {}", pattern.clone()}.as_str())
+                glob(pattern).expect(format! {"Invalid file pattern: {}", pattern.clone()}.as_str())
```
[the remark this answers is not in the document]
Comment (soldni) [reply src/shard.rs:606]: I'm using `make style`, which runs `rustfmt --edition 2021 src/*.rs`. I've also added  a style check as part of the CI

File: src/s3_util.rs [hunk 3]
```diff
@@ -135,7 +135,12 @@ pub fn find_objects_matching_patterns(
         let mut has_more = true;
         let mut token: Option<String> = None;
         while has_more {
-            let (bucket, key) = split_url(&prefix).unwrap();
+            let (bucket, key) = match split_url(&prefix) {
```
[the remark this answers is not in the document]
Comment (soldni) [reply src/s3_util.rs:138]: 👍 making a note for next PR

File: src/deduper.rs [hunk 4]
```diff
@@ -16,25 +15,23 @@ use threadpool::ThreadPool;
...
 
-pub fn run(config: DeduperConfig) {
-    assert!(
-        config.dedupe.paragraphs.is_some() ^ config.dedupe.documents.is_some(),
-        "Must dedupe either paragraphs or documents"
-    );
-
-    let s3_client = s3_util::new_client(None).unwrap();
-
+pub fn run(config: DeduperConfig) -> Result<u32, u32> {
```
[the remark this answers is not in the document]
Comment (soldni) [reply src/deduper.rs:22]: oooooh yeah

**Pull request merged**